pinot-plugins/pinot-stream-ingestion/pinot-kinesis/src/main/java/org/apache/pinot/plugin/stream/kinesis/KinesisPartitionGroupOffset.java

```
@@ -0,0 +1,94 @@
+import org.apache.pinot.spi.utils.JsonUtils;
+
+
+/**
+ * A {@link StreamPartitionMsgOffset} implementation for the Kinesis partition group consumption
+ * A partition group consists of 1 or more shards. The KinesisCheckpoint maintains a Map of shards to the
+ * sequenceNumber.
+ * The sequenceNumber is the id (equivalent to offset in kafka) for the messages in the shard.
+ * From the Kinesis documentation:
+ * Each data record has a sequence number that is unique per partition-key within its shard.
+ * Kinesis Data Streams assigns the sequence number after you write to the stream with client.putRecords or client
+ * .putRecord.
+ * Sequence numbers for the same partition key generally increase over time.
+ * The longer the time period between write requests, the larger the sequence numbers become.
+ */
+public class KinesisPartitionGroupOffset implements StreamPartitionMsgOffset {
+  private final Map<String, String> _shardToStartSequenceMap;
+
+  public KinesisPartitionGroupOffset(Map<String, String> shardToStartSequenceMap) {
+    _shardToStartSequenceMap = shardToStartSequenceMap;
+  }
+
+  public KinesisPartitionGroupOffset(String offsetStr)
+      throws IOException {
+    _shardToStartSequenceMap = JsonUtils.stringToObject(offsetStr, new TypeReference<Map<String, String>>() {
+    });
+  }
+
+  public Map<String, String> getShardToStartSequenceMap() {
+    return _shardToStartSequenceMap;
+  }
+
+  @Override
+  public String toString() {
+    try {
+      return JsonUtils.objectToString(_shardToStartSequenceMap);
+    } catch (JsonProcessingException e) {
+      throw new IllegalStateException(
+          "Caught exception when converting KinesisCheckpoint to string: " + _shardToStartSequenceMap);
+    }
+  }
+
+  @Override
+  public KinesisPartitionGroupOffset fromString(String kinesisCheckpointStr) {
+    try {
+      return new KinesisPartitionGroupOffset(kinesisCheckpointStr);
+    } catch (IOException e) {
```

Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)